### PR TITLE
Start process instance at specific element

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This PR will make it possible to start a PI anywhere in the process. It will look in the start instructions on which element(s) the process should be started. It will start by activating the process itself. Using recursing it will then activate all the flow scopes (e.g. subprocesses) that are required for the element in the start instruction. It will do this in the correct order (top to bottom). Lastly it will send an activate command for the element(s) in the start instructions.

## Related issues

closes #9390 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
Requesting a review from both of you as it is important to do some knowledge sharing here.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:96]: :wrench: You could pass in just the start instructions. The record is only used to get to the start instructions

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:242]: :x: I think this could activate an element twice. E.g. given this process:
image

With  `client.createProcessInstance().startBefore("task").startBefore("subprocess").send().join()` the sub process would be activated twice. The first time as part of `activateFlowScopes()` for `task` and the second time when the second start instruction is processed.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:256]:
```suggestion
   * activated here as an activate command is sent for this later on.
```

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java
Comment (pihme) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/ElementInstance.java:201]: :thinking: Do not understand this change

**Review by pihme — CHANGES_REQUESTED** [review]
Good work, but I do have some concerns in the details.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:223]: Or maybe I am wrong. I routinely get confused which bit is done where.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:223]: I believe the variable behaviour is part of this Processor. One of the first things that happens when receiving this command is merging the variables from the command into the scope of the process instance.
```java
    final long processInstanceKey = keyGenerator.nextKey();
    if (!setVariablesFromDocument(
        controller, record, process.getKey(), processInstanceKey, process.getBpmnProcessId())) {
      return true;
    }
```
```java
  private boolean setVariablesFromDocument(
      final CommandControl<ProcessInstanceCreationRecord> controller,
      final ProcessInstanceCreationRecord record,
      final long processDefinitionKey,
      final long processInstanceKey,
      final DirectBuffer bpmnProcessId) {
    try {
      variableBehavior.mergeLocalDocument(
          processInstanceKey,
          processDefinitionKey,
          processInstanceKey,
          bpmnProcessId,
          record.getVariablesBuffer());
    } catch (final MsgpackReaderException e) {
      Loggers.PROCESS_PROCESSOR_LOGGER.error(ERROR_INVALID_VARIABLES_LOGGED_MESSAGE, e);
      controller.reject(
          RejectionType.INVALID_ARGUMENT,
          String.format(ERROR_INVALID_VARIABLES_REJECTION_MESSAGE, e.getMessage()));

      return false;
    }

    return true;
  }
```

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:223]: I do think a test to verify this is warranted so I will add this.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java]: I think you are right. It's a bit of an odd scenario. What would the expected behaviour be in this case? 
Send a command to activate the subprocess, and then send a command to activate the task. This would sort of create 2 process instances in one 😅 Then the task would be activated twice.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:77]: :wrench: We could add a static import for `tuple()` to reduce the code :) 

```suggestion
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATING),
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
            tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATING),
            tuple(BpmnElementType.END_EVENT, ProcessInstanceIntent.ELEMENT_ACTIVATED));
```

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:417]: :wrench: We could limit the stream to the process instance instead to verify that the variable is created only for the process instance and not more. For example:

```
RecordingExporter.
  records()
  .limitToProcessInstance(key)
  .variableRecords()
  // ...
```

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java]: It is valid user input though. 

Could be a misunderstanding. Maybe the customer didn't know we will activate the surrounding event scopes and so they want to do it explicitly

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:27]: :wrench: We should add another test to verify the properties of the `activate` command and the `activating` + `activated` events that are written by the processor.

If the processor writes the records not correctly then it can be hard to root cause the issue. For example, a wrong flow scope key caused the process to not complete.

**Review by saig0 — APPROVED** [review]
@remcowesterhoud good job. I like it :+1: 

I have only a few minor suggestions. Nothing major. 

Don't worry If you don't have time to finish the PR today. I can take over :handshake:

File: engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java
Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java:75]: I don't agree with this. The record we are creating contains a `startInstruction` property. It makes perfect sense to me to have a `withStartInstruction` method which will add this instruction. Whether it is start before or after an element is irrelevant. Especially since there is only one possibility. If we ever decide to support multiple options they should be added to the start instruction that is passed.

I don't like to specify it in the test names either. We don't know the concept of starting before or after an element in Zeebe, so why add it to the test name? I could understand doing it in the Java client for backwards compatibility down the road, but here I feel it's unnecessary.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:417]: A limit like this doesn't work as `limitToProcessInstance` also filters on specific `ProcessInstanceIntents`, thus not returning the variable records. I don't see another way to limit, am I missing something?

File: engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/util/client/ProcessInstanceClient.java:75]: Okay :smile: I don't force you to change.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:406]: Not true. It limits the stream based on the process instance records of the given process instance key but it returns the variable records.

However, I just realized that it doesn't work because the variables are written before the process instance record of the process is written. 

But we could write it in the following way:

```
Assertions.assertThat(
            RecordingExporter.records()
                .limit(
                    // extract the condition into a method like limitToProcessInstance()
                    r ->
                        r.getKey() == key
                            && r.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETED)
                .variableRecords())
        .hasSize(1)
        .extracting(Record::getValue)
        .extracting(
            VariableRecordValue::getScopeKey,
            VariableRecordValue::getName,
            VariableRecordValue::getValue)
        .containsExactly(tuple(key, "variable", "\"value\""));
```

**Comment by remcowesterhoud** [comment]
I have addressed most of your comments. The only remaining things are the unresolved comments. They should be fairly easy to implement. @saig0 I will leave it in your hands 😄

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java]: > With client.createProcessInstance().startBefore("task").startBefore("subprocess").send().join() the sub process would be activated twice. The first time as part of activateFlowScopes() for task and the second time when the second start instruction is processed.

Yes, it would create the subprocess twice. For me, this is the expected behavior. 

The rule for process instance modification is: it creates a new element instance in the existing scope instance. If no scope instance exists yet then a new scope instance is created.

The start-process-anywhere command follows the same rule. 

Independent of the order of the two instructions, the scope instance of the subprocess doesn't exist yet. As a result, it will create a new scope instance of the subprocess twice.

I'll create a new test case for it.

**Comment by saig0** [comment]
bors merge

**Comment by saig0** [comment]
@pihme please have another look. I need your approval to merge the PR :heavy_check_mark:

File: broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java
Comment (pihme) [inline broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java:71]: :x: I find this naming confusing. Maybe this could be rewritten to
`betweenProcessInstanceEvents(processInstanceKey, List.of(ProcessInstanceIntent.ELEMENT_ACTIVATING), List.of(ProcessInstanceEvent.ELEMENT_COMPLETED, ProcessInstanceEvent.ELEMENT_TERMINATED))`

Although _events_ is also ambiguous. I am looking for a word that describes a point in the lifecycle of a process instance.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (pihme) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:382]: :question: I don't understand how this works, to be honest. The `limitToProcessInstance` gives me all records where the key is the process instance (and before it is being closed). I would assume this predicate is positive for all process instance records. But somehow there seem to be variable records among them, or else this test doesn't work.
Are the variable records which have as key the process instance??? Because the key on a variable records points to a scope and the key on a process instance record is the key of the process instance?

Overall, I find this difficult to comprehend. This test is about the variables that are being written. Sp is there no way to rewrite this as:

`
RecordingExporter.variableRecords().withEventScopeKey(processInstanceKey).withIntent(VariableIntent.CREATED).findFirst()
`

**Review by pihme — CHANGES_REQUESTED** [review]
Almost there. I still have some nitpicky comments/questions and one change request.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java:382]: >  I don't understand how this works, to be honest. The limitToProcessInstance gives me all records where the key is the process instance

No, your assumption is not correct. The method uses the `limit()` function of the stream. It returns all records (of all types) until the `completed` event of the process instance.

> is there no way to rewrite this as:

Sure, we could write the condition in this way. But this doesn't verify that any other variable not written, for example, for another scope.

File: broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java
Comment (saig0) [reply broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java:71]: I accept your objections regarding the naming but I don't see a reason to block this PR. We can change the name later if we have a better idea.

Comment (pihme) [reply broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java:71]: Then let's create an issue for the renaming, so we don't forget it. And then this PR can go ahead.

Comment (saig0) [reply broker/src/test/java/io/camunda/zeebe/broker/engine/RecordVersionTest.java:71]: I will not create an issue for the renaming. If someone has a better idea for the name then we can just create a PR. I don't see that this name is so important.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/ProcessEventProcessors.java
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/ProcessEventProcessors.java]: I extracted the metrics and reuse them in both places.

**Comment by saig0** [comment]
@pihme I addressed your comments. Let's approve the PR.

**Review by pihme — APPROVED** [review]
bors merge

**Comment by saig0** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by saig0** [comment]
The build failed repeatably with: 

```
Resolved Java 17.0.3+7 from tool-cache
Setting Java 17.0.3+7 as the default
Java configuration:
  Distribution: temurin
  Version: [17](https://github.com/camunda/zeebe/runs/6859178109?check_suite_focus=true#step:3:18).0.3+7
  Path: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/17.0.3-7/x64
Creating settings.xml with server-id: github
Writing to /home/runner/.m2/settings.xml
Error: getCacheEntry failed: Cache service responded with 503
```

**Comment by saig0** [comment]
bors merge

**Comment by pihme** [comment]
bors merge

**Pull request merged**
